components/Table.tsx

```diff
@@ -87,6 +86,91 @@ export const Avatar = styled.img`
   width: 40px;
 `;
 
+function LocationFilter({ column: { filterValue, setFilter, preFilteredRows, id } }) {
+  const options = React.useMemo(() => {
+    const options = [];
+    const regions = [];
+    const countries = [];
+    const domesticRegions = [];
+    preFilteredRows.forEach(row => {
+      if (row.values.location.region && !regions.includes(row.values.location.region)) {
+        regions.push(row.values.location.region);
+      }
+
+      if (row.values.location.countryCode && !countries.find(c => c.countryCode === row.values.location.countryCode)) {
+        countries.push({ countryCode: row.values.location.countryCode, region: row.values.location.region });
+      }
+      if (
+        row.values.location.domesticRegion &&
+        !domesticRegions.find(c => c.domesticRegion === row.values.location.domesticRegion)
+      ) {
+        domesticRegions.push({
+          domesticRegion: row.values.location.domesticRegion,
+          countryCode: row.values.location.countryCode,
+        });
+      }
+    });
+    const sortedRegions = regions.sort();
+
+    // for each region
+    sortedRegions.forEach(region => {
+      // add the region to the options
+      options.push({ type: 'region', label: `${region}`, value: region });
+      // add the countries in that region to the options
+      countries
+        .filter(country => country.region === region)
+        .map(country => country.countryCode)
+        .sort()
+        .forEach(country => {
+          options.push({ type: 'countryCode', label: `- ${country}`, value: country });
+
+          // add the domestic regions in that country to the options
+          domesticRegions
+            .filter(domesticRegion => domesticRegion.countryCode === country)
+            .map(state => state.domesticRegion)
+            .sort()
+            .forEach(domesticRegion => {
+              options.push({ type: 'domesticRegion', label: `-- ${domesticRegion}`, value: domesticRegion });
+            });
+        });
+    });
+
+    return options;
+  }, [id, preFilteredRows]);
```

Comment:
:ice_cream: This whole thing is difficult to read, would be nice to extract it in a separate function. Something like:

```jsx
const options = React.useMemo(() => getOptions(preFilteredRows), [preFilteredRows]);
```

Reply:
@Betree Good call, pushed a fix, thank you!